syntax = "proto3";

import "proto/db.proto";

package api;

message EmptyReq {}
message EmptyResp {}
message ErrorResp {
    string err_msg = 1;
}

// --- all question ---
message CreateAllQuestionReq {
    db.Question.Type type = 1;
    string description = 2;
    db.JudgeData judge_data = 3;
    db.ChoiceData choice_data = 4;
}

message CreateAllQuestionResp {
    db.Question question = 1;
}

message ListAllQuestionReq {
    message Filters {
        int32 q_id = 1;
        string description = 2;
        db.Question.Type type = 3;
    }
    int32 current = 1;
    int32 page_size = 2;
    Filters filters = 3;
}

message ListAllQuestionResp {
    int32 total = 1;
    repeated db.Question data = 2;
}

// --- exam ---
message CreateExamPaperReq {
    string title = 1;
    repeated int32 q_id_list = 2;
}

message CreateExamPaperResp {
    db.ExamPaper exam_paper = 1;
}

message GetExamPaperReq {
    int32 e_id = 1;
}

message GetExamPaperResp {
    db.ExamPaper exam_paper = 1;
    repeated db.Question questions = 2;
}

message ListExamPaperReq {
    message Filters {
        int32 e_id = 1;
        string title = 2;
    }
    int32 current = 1;
    int32 page_size = 2;
    Filters filters = 3;
}

message ListExamPaperResp {
    int32 total = 1;
    repeated db.ExamPaper data = 2;
}

message LoginReq {
    int64 uid = 1;
    string password = 2;
}

message LoginResp {
    db.User user = 1;
}

message RegisterReq {
    int64 uid = 1;
    string name = 2;
    string password = 3;
}

message RegisterResp {
    db.User user = 1;
}

message GetUserInfoReq {
    int64 uid = 1;
}

message GetUserInfoResp {
    db.User user = 1;
}

message fetchCurrentUserResp {
    string name = 1;
}